### PR TITLE
New mesh refinement strategies

### PR DESCRIPTION
**Pull request by tjhei**
- new mesh refinement based on artifical viscosity
- new mesh refinement based on composition gradient or threshold

Note: This needs a rebase once #853 is merged.

### CODE REVIEW

File: include/aspect/mesh_refinement/composition_threshold.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#include <aspect/simulator_access.h>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+
+    /**
+     * A class that implements a mesh refinement criterion based on the
+     * absolute value of the compositional fields. Every cell than contains a
```
Comment (bangerth) [inline include/aspect/mesh_refinement/composition_threshold.h]: than -> that

File: source/mesh_refinement/artificial_viscosity.cc [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+          prm.declare_entry("Temperature scaling factor",
+                            "0.0",
+                            Patterns::Double(0),
+                            "A scaling factor for the artificial viscosity "
+                            " of the temperature equation. Use 0.0 to disable.");
+          prm.declare_entry("Compositional field scaling factors",
+                            "",
+                            Patterns::List (Patterns::Double(0)),
+                            "A list of scaling factors by which every individual compositional "
+                            "field will be multiplied by. These "
```
Comment (bangerth) [inline source/mesh_refinement/artificial_viscosity.cc]: drop the second "by"

File: source/mesh_refinement/artificial_viscosity.cc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                       ExcMessage ("The number of scaling factors given here must either be "
+                                   "zero or equal to the number of chosen refinement criteria."));
+
+          if (composition_scaling_factors.size() == 0)
+            composition_scaling_factors.resize (this->n_compositional_fields(), 0.0);
+
+          double sum_compositional_factors = 0.0;
+          for (unsigned int i=0; i<composition_scaling_factors.size(); ++i)
+            {
+              sum_compositional_factors += composition_scaling_factors[i];
```
Comment (bangerth) [inline source/mesh_refinement/artificial_viscosity.cc]: `Assert (temperature_scaling_factor + std::accumulate (comp....begin(), comp....end(), 0.0) > 0, ExcMessage(...));`

File: source/mesh_refinement/composition_gradient.cc [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+          unsigned int i=0;
+          for (; cell!=endc; ++cell, ++i)
+            if (cell->is_locally_owned())
+              {
+                fe_values.reinit(cell);
+                fe_values[this->introspection().extractors.compositional_fields[c]].get_function_gradients (this->get_solution(),
+                    composition_gradients);
+
+                // for each composition dof, write into the output
+                // vector the compositon gradient. Note that quadrature points and
```
Comment (bangerth) [inline source/mesh_refinement/composition_gradient.cc]: typo

File: source/mesh_refinement/composition_gradient.cc [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Composition gradient");
+        {
+          prm.declare_entry("Compositional field scaling factors",
+                            "",
+                            Patterns::List (Patterns::Double(0)),
+                            "A list of scaling factors by which every individual compositional "
+                            "field gradient will be multiplied by. If only a single compositional "
```
Comment (bangerth) [inline source/mesh_refinement/composition_gradient.cc]: remove second "by"

File: source/mesh_refinement/composition_gradient.cc [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+{
+  namespace MeshRefinement
+  {
+    ASPECT_REGISTER_MESH_REFINEMENT_CRITERION(CompositionGradient,
+                                              "composition gradient",
+                                              "A mesh refinement criterion that computes refinement "
+                                              "indicators from the gradients of compositional fields. "
+                                              "If there is more than one compositional field, then "
+                                              "it simply takes the sum of the indicators computed "
+                                              "from each of the compositional field.")
```
Comment (bangerth) [inline source/mesh_refinement/composition_gradient.cc]: ...times a user-specified weight for each field.

**Comment by bangerth** [comment]
I'm not sure what the purpose is to refine based on the artificial viscosity, but there is nothing wrong with it either.

Please rebase, address the comments I sent, and merge.

**Comment by tjhei** [comment]
> I'm not sure what the purpose is to refine based on the artificial viscosity, but there is nothing wrong with it either.

I am not sure either. :-) We were comparing many different strategies and it sounded at least plausible to look at this because it represents the entropy (or local residual) of the system.

**Comment by tjhei** [comment]
rebased and good to go.

**Pull request merged**